File: src/leetCodeProblems/MathCalculations/RomanToInteger13.java
package leetCodeProblems.MathCalculations;

public class RomanToInteger13 {

	public int romanToInt(String s) {
        
        int answerInt = 0;
        
        for (int i=0; i < s.length(); i++) {
            
            boolean twoCharPairsMatched = false;
            
            if (i+1 < s.length()) {
                
                StringBuilder sb = new StringBuilder();
                sb.append(s.charAt(i));
                sb.append(s.charAt(i+1));
                
                String twoCharsPairs = sb.toString();
                
                switch (twoCharsPairs) {
                    case "CM":
                        answerInt += 900;
                        twoCharPairsMatched = true;
                        break;
                    case "CD":
                        answerInt += 400;
                        twoCharPairsMatched = true;
                        break;
                    case "XC":
                        answerInt += 90;
                        twoCharPairsMatched = true;
                        break;
                    case "XL":
                        answerInt += 40;
                        twoCharPairsMatched = true;
                        break;
                    case "IX":
                        answerInt += 9;
                        twoCharPairsMatched = true;
                        break;
                    case "IV":
                        answerInt += 4;
                        twoCharPairsMatched = true;
                        break;
                    default:
                        //Nothing
                        
                }
            }
            
            if (twoCharPairsMatched) {
                i++;
            }
            else {
                switch(s.charAt(i)) {
                    case 'M':
                        answerInt += 1000;
                        break;
                    case 'D':
                        answerInt += 500;
                        break;
                    case 'C':
                       answerInt += 100;
                       break;
                    case 'L':
                        answerInt += 50;
                        break;
                    case 'X':
                        answerInt += 10;
                        break;
                    case 'V':
                       answerInt += 5;
                       break;
                    case 'I':
                       answerInt += 1;
                       break;
                    default:
                        //Nothing
               } 
            }
        }
        
        return answerInt;
        
    }
	
	public static void main(String[] args) {
		
		String roman = "MCMXCIV";
		
		RomanToInteger13 obj = new RomanToInteger13();
		
		System.out.println(obj.romanToInt(roman));
	}
}

File: src/leetCodeProblems/MathCalculations/RomanToInteger13.go
package main

/*
- Leetcode - https://leetcode.com/problems/roman-to-integer/
- Time - O(n)
- Space - O(1)
*/
func romanToInt(s string) int {
	out := 0

	for i := 0; i < len(s); i++ {
		switch string(s[i]) {
		case "M":
			out += 1000
		case "D":
			out += 500
		case "C":
			if i+1 < len(s) {
				switch string(s[i+1]) {
				case "D":
					out += 400
					i++
					continue
				case "M":
					out += 900
					i++
					continue
				}
			}
			out += 100
		case "L":
			out += 50
		case "X":
			if i+1 < len(s) {
				switch string(s[i+1]) {
				case "L":
					out += 40
					i++
					continue
				case "C":
					out += 90
					i++
					continue
				}
			}
			out += 10
		case "V":
			out += 5
		case "I":
			if i+1 < len(s) {
				switch string(s[i+1]) {
				case "V":
					out += 4
					i++
					continue
				case "X":
					out += 9
					i++
					continue
				}
			}
			out += 1
		}
	}
	return out
}

// func main() {
// 	log.Println(romanToInt("III"))
// 	log.Println(romanToInt("LVIII"))
// 	log.Println(romanToInt("MCMXCIV"))
// }
